

// [START declaration]
syntax = "proto3";
// Namespace
package tutorial;
// Protobuf supports import other proto files
import "google/protobuf/timestamp.proto";
// [END declaration]


// Message hierarchy
// Addressbook -> Person -> PhoneNumber

// [START messages]
// We create a Person message

message Person {
    string name = 1;
    int32 id = 2;
    string email = 3;

    // Protobuf support enumerator type
    enum PhoneType {
        MOBILE = 0;  // Default is always 0 in proto3
        HOME = 1;
        WORK = 2;
    }

    // Inside the Person struct we have PhoneNumber message
    message PhoneNumber {
        string number = 1;
        PhoneType type = 2;
    }

    // Each element in a repeated field requires re-encoding the tag number
    repeated PhoneNumber phones = 4;

    google.protobuf.Timestamp last_updated = 5;
}

// Address book file is just one of these.
message AddressBook {
    repeated Person people = 1;
}

// [END messages]
